replica/database.cc

```diff
@@ -1808,6 +1809,14 @@ future<> database::apply_in_memory(const frozen_mutation& m, schema_ptr m_schema
 
     data_listeners().on_write(m_schema, m);
 
+    if (m.representation().size() > 128*1024) {
+        return unfreeze_gently(m, std::move(m_schema)).then([&cf, h = std::move(h), timeout] (auto m) mutable {
+            return do_with(std::move(m), [&cf, h = std::move(h), timeout] (auto& m) mutable {
+                return cf.apply(m, std::move(h), timeout);
+            });
+        });
+    }
```

Comment:
will apply(mutation) yield? I'm guessing not, but I'm okay with leaving it for later.

Reply:
> Since there's a do_with here, how about extracting it to a coroutine method?

That's exactly what I did in the previous version, bit @denesb had reservations about that since the database class has a proliferation of methods already so he suggested to inline it.

> will apply(mutation) yield? I'm guessing not, but I'm okay with leaving it for later.

It might, since it's waiting for available memory, and the caller must keep the mutation alive.
```
    return dirty_memory_region_group().run_when_memory_available([this, &m, m_schema = std::move(m_schema), h = std::move(h), &cg, holder = std::move(holder)]() mutable {
        do_apply(cg, std::move(h), m, m_schema);
    }, timeout);
```

Reply:
> > Since there's a do_with here, how about extracting it to a coroutine method?
> 
> That's exactly what I did in the previous version, bit @denesb had reservations about that since the database class has a proliferation of methods already so he suggested to inline it.

Coroutinization is there to increase readability, and it can't be inlined since that makes the caller a coroutine too.


> 
> There was no difference in performance either way.

Of course, since you measure <128k writes.


> 
> > will apply(mutation) yield? I'm guessing not, but I'm okay with leaving it for later.
> 
> It might, since it's waiting for available memory, and the caller must keep the mutation alive.
> 

I meant, yield while applying a large mutation. It will probably stall.


> ```
>     return dirty_memory_region_group().run_when_memory_available([this, &m, m_schema = std::move(m_schema), h = std::move(h), &cg, holder = std::move(holder)]() mutable {
>         do_apply(cg, std::move(h), m, m_schema);
>     }, timeout);
> ```

Reply:
> > > Since there's a do_with here, how about extracting it to a coroutine method?
> > 
> > 
> > That's exactly what I did in the previous version, bit @denesb had reservations about that since the database class has a proliferation of methods already so he suggested to inline it.
> 
> Coroutinization is there to increase readability, and it can't be inlined since that makes the caller a coroutine too.

There's a hybrid option as well, but I'm not sure it's worth it.
I'm ok wither way, so if you prefer a coroutine method, I'll add it back.
The hybrid option that saves the `do_with` is:
```
        return unfreeze_gently(m, std::move(m_schema)).then(coroutine::lambda([&cf, h = std::move(h), timeout] (auto m) mutable -> future<> {
            co_await cf.apply(m, std::move(h), timeout);
        }));
```

> 
> > There was no difference in performance either way.
> 
> Of course, since you measure <128k writes.

I commented out the other options and measured only this part - inline continuations vs. coroutine method.

> > > will apply(mutation) yield? I'm guessing not, but I'm okay with leaving it for later.
> > 
> > 
> > It might, since it's waiting for available memory, and the caller must keep the mutation alive.
> 
> I meant, yield while applying a large mutation. It will probably stall.

maybe I misunderstand you, but applying of the mutation is synchronous in table::do_apply.
The only place that table::apply may yield is `run_when_memory_available`

> > ```
> >     return dirty_memory_region_group().run_when_memory_available([this, &m, m_schema = std::move(m_schema), h = std::move(h), &cg, holder = std::move(holder)]() mutable {
> >         do_apply(cg, std::move(h), m, m_schema);
> >     }, timeout);
> > ```

Reply:
Mixed continuations/coroutines are a great way to obfuscate the code.